### PR TITLE
add meta op select

### PR DESCRIPTION
**Pull request by SunNy820828449**
this pr is for add meta op select register.
select op has three input: condition(type == bool),  true_value,  false_value.
for each element in tensor:
  if(condition[i])
    output[i] = true_value[i];
  else
    output[i] = false_value[i];

image

### CODE REVIEW

File: cinn/hlir/op/nn.cc [hunk 1]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
 
+std::shared_ptr<OpStrategy> StrategyForSelect(const framework::NodeAttr &attrs,
+                                              const std::vector<ir::Tensor> &inputs,
+                                              const std::vector<Type> &out_type,
+                                              const std::vector<std::vector<int>> &output_shapes,
+                                              const Target &target) {
+  framework::CINNCompute select_compute([=](lang::Args args, lang::RetValue *ret) {
+    CHECK(!args.empty()) << "The input argument of select compute is empty! Please check.\n";
+    CINNValuePack arg = args[0];
+    CHECK(arg.size() >= 3) << "at least one input tensor for select compute\n";
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc]: CHECK_GE(), at least 3 input tensors

File: cinn/hlir/op/nn.cc [hunk 2]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+    Expr true_value  = arg[1];
+    Expr false_value = arg[2];
+    CHECK(condition.as_tensor());
+    CHECK(true_value.as_tensor());
+    CHECK(false_value.as_tensor());
+    auto out = pe::Select(
+        condition.as_tensor_ref(), true_value.as_tensor_ref(), false_value.as_tensor_ref(), UniqName("Select_output"));
+    auto stages = CreateStages({out});
+    *ret        = CINNValuePack{{CINNValue(Expr(out.get())), CINNValue(stages)}};
+  });
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc:1816]: CINNValue(out)

File: cinn/hlir/op/nn.cc [hunk 3]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+
+  framework::CINNSchedule select_schedule([=](lang::Args args, lang::RetValue *ret) {
+    CHECK(!args.empty()) << "The input argument of select schedule is empty! Please check.\n";
+    CINNValuePack arg_pack = args[0];
+    CHECK_EQ(arg_pack.size(), 2UL);
+    if (target.arch == Target::Arch::NVGPU) {
+      Expr out              = arg_pack[0];
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc]: 公共代码提出去

File: cinn/hlir/op/nn.cc [hunk 4]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+      poly::StageMap stages = arg_pack[1];
+      CHECK(out.as_tensor());
+      pe::CudaScheduleInjective(stages[out.as_tensor_ref()], output_shapes.back(), target);
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc]: output_shapes.front(),下同

File: cinn/hlir/op/nn.cc [hunk 5]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+  }
+  return strategy;
+}
+
+std::vector<framework::shape_t> InferShapeForSelect(const std::vector<framework::shape_t> &inputs_shape,
+                                                    framework::NodeAttr &attrs,
+                                                    const Target &target) {
+  CHECK(!inputs_shape.empty() && !inputs_shape[0].empty()) << "The input's shape size is 0! Please check again.";
+  std::vector<framework::shape_t> res{inputs_shape[0]};
+  return res;
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc:1853]: check输入输出shape一致

File: cinn/hlir/op/nn.cc [hunk 6]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+}
+
+std::vector<Type> InferDtypeForSelect(const std::vector<Type> &inputs_type,
+                                      const framework::NodeAttr &attrs,
+                                      const Target &target) {
+  CHECK(!inputs_type.empty()) << "The input's type size is 0! Please check again.";
+  std::vector<Type> res{inputs_type[1]};
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc:1861]: 避免数组越界，check inputs_type大于1

File: cinn/hlir/pe/nn.cc [hunk 7]
```diff
@@ -940,6 +940,18 @@ Tensor DropoutInfer(const ir::Tensor &tensor,
   }
 }
 
+ir::Tensor Select(const ir::Tensor &condition,
+                  const ir::Tensor &true_value,
+                  const ir::Tensor &false_value,
+                  const std::string &output_name) {
+  return lang::Compute(
+      condition->shape,
+      [=](const std::vector<Expr> &indice) {
```
Comment (wenming2014) [inline cinn/hlir/pe/nn.cc:1018]: check 输入输出shape一致

File: cinn/lang/builtin.h [hunk 8]
```diff
@@ -120,6 +120,10 @@ inline Expr ReduceMin(Expr e, const std::vector<Var>& reduce_axis, Expr initial
   return ir::Reduce::Make(ir::Reduce::kMin, initial, e, reduce_axis);
 }
 
+inline Expr Select(Expr condition, Expr true_value, Expr false_value) {
+  return ir::Select::Make(condition, true_value, false_value);
```
Comment (wenming2014) [inline cinn/lang/builtin.h]: 这个库上应该有接口, cinn/common/ir_util.h

File: cinn/hlir/pe/nn.cc [hunk 9]
```diff
@@ -940,6 +940,18 @@ Tensor DropoutInfer(const ir::Tensor &tensor,
   }
 }
 
+ir::Tensor Select(const ir::Tensor &condition,
+                  const ir::Tensor &true_value,
+                  const ir::Tensor &false_value,
+                  const std::string &output_name) {
+  return lang::Compute(
```
Comment (wenming2014) [inline cinn/hlir/pe/nn.cc:1016]: check condition type

File: cinn/hlir/op/nn.cc [hunk 10]
```diff
@@ -1827,6 +1827,71 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+  } else {
+    LOG(FATAL) << "Select op with dtype != float32 is not implemented yet!";
+  }
+  return strategy;
+}
+
+std::vector<framework::shape_t> InferShapeForSelect(const std::vector<framework::shape_t> &inputs_shape,
+                                                    framework::NodeAttr &attrs,
+                                                    const Target &target) {
+  CHECK(!inputs_shape.empty() && !inputs_shape[0].empty()) << "The input's shape size is 0! Please check again.";
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc]: check condition type, 组网的时候验证合法性

Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc]: done, thanks

File: cinn/hlir/pe/nn.cc (hunk 9 above)
Comment (SunNy820828449) [reply cinn/hlir/pe/nn.cc:1016]: done thanks

File: cinn/lang/builtin.h (hunk 8 above)
Comment (SunNy820828449) [reply cinn/lang/builtin.h]: 需要使用封装的接口吗？还是保留使用make的接口

File: cinn/hlir/op/nn.cc (hunk 6 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc:1861]: done,thanks

File: cinn/hlir/op/nn.cc (hunk 4 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc]: done,thanks

File: cinn/hlir/op/nn.cc (hunk 3 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc]: done,thanks

File: cinn/hlir/op/nn.cc (hunk 2 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc:1816]: done,thanks

File: cinn/hlir/op/nn.cc (hunk 1 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc]: done,thanks

File: cinn/hlir/op/nn.cc (hunk 5 above)
Comment (SunNy820828449) [reply cinn/hlir/op/nn.cc:1853]: done， thanks

File: cinn/hlir/pe/nn.cc (hunk 7 above)
Comment (SunNy820828449) [reply cinn/hlir/pe/nn.cc:1018]: done,thanks

File: cinn/hlir/op/nn.cc [hunk 11]
```diff
@@ -1827,6 +1827,74 @@ std::vector<Type> InferDtypeForDropoutInfer(const std::vector<Type> &inputs_type
...
+  CHECK(inputs_shape[0] == inputs_shape[1] && inputs_shape[1] == inputs_shape[2])
+      << "input tensor shapes is not equal!";
+  std::vector<framework::shape_t> res{inputs_shape[0]};
+  return res;
+}
+
+std::vector<Type> InferDtypeForSelect(const std::vector<Type> &inputs_type,
+                                      const framework::NodeAttr &attrs,
+                                      const Target &target) {
+  CHECK_GT(inputs_type.size(), 3) << "The input's type size is 0! Please check again.";
```
Comment (wenming2014) [inline cinn/hlir/op/nn.cc]: CHECK_GE吧，The input's type size is 0修改下？

File: cinn/hlir/op/op_nn_test.cc [hunk 12]
```diff
@@ -304,6 +304,64 @@ TEST(Operator, Operator_Pool1d_Test0) {
...
+  cinn_buffer_t *A_buf = common::BufferBuilder(Bool(), {16, 64, 64}).set_random().Build();
+  cinn_buffer_t *B_buf = common::BufferBuilder(Float(32), {16, 64, 64}).set_random().Build();
+  cinn_buffer_t *C_buf = common::BufferBuilder(Float(32), {16, 64, 64}).set_random().Build();
+  cinn_buffer_t *D_buf = common::BufferBuilder(Float(32), {16, 64, 64}).set_random().Build();
+  cinn_pod_value_t a_arg(A_buf), b_arg(B_buf), c_arg(C_buf), d_arg(D_buf);
+  cinn_pod_value_t args[] = {a_arg, b_arg, c_arg, d_arg};
+  fn_(args, 4);
+
+  ASSERT_EQ(impl->name, "strategy.select.x86");
+  ASSERT_EQ(select->description, "This operator implements the meta op 'Select'.");
```
Comment (wenming2014) [inline cinn/hlir/op/op_nn_test.cc:387]: 测试下精度？

File: cinn/lang/builtin.h (hunk 8 above)
Comment (wenming2014) [reply cinn/lang/builtin.h]: 之前应该有封装的接口，在cinn/common/ir_util.h里

Comment (SunNy820828449) [reply cinn/lang/builtin.h]: done,thanks

File: cinn/backends/llvm/execution_engine.cc [hunk 13]
```diff
@@ -146,6 +146,8 @@ void ExecutionEngine::Link(const ir::Module &module) {
   VLOG(3) << "ir_emitter->Compile(module) Succeed!";
   CHECK(!llvm::verifyModule(*m, &llvm::errs())) << "Invalid module found";
 
+  // m->dump();
```
Comment (wenming2014) [inline cinn/backends/llvm/execution_engine.cc]: 删除注释

**Review by wenming2014 — DISMISSED** [review]
LGTM

**Review by wenming2014 — APPROVED** [review]

**Pull request merged**